packages/maskbook/src/plugins/ITO/hooks/useRegion.ts

```diff
@@ -0,0 +1,312 @@
+    | 'GY'
+    | 'HT'
+    | 'HM'
+    | 'VA'
+    | 'HN'
+    | 'HK'
+    | 'HU'
+    | 'IS'
+    | 'IN'
+    | 'ID'
+    | 'IR'
+    | 'IQ'
+    | 'IE'
+    | 'IM'
+    | 'IL'
+    | 'IT'
+    | 'JM'
+    | 'JP'
+    | 'JE'
+    | 'JO'
+    | 'KZ'
+    | 'KE'
+    | 'KI'
+    | 'KP'
+    | 'KR'
+    | 'KW'
+    | 'KG'
+    | 'LA'
+    | 'LV'
+    | 'LB'
+    | 'LS'
+    | 'LR'
+    | 'LY'
+    | 'LI'
+    | 'LT'
+    | 'LU'
+    | 'MO'
+    | 'MK'
+    | 'MG'
+    | 'MW'
+    | 'MY'
+    | 'MV'
+    | 'ML'
+    | 'MT'
+    | 'MH'
+    | 'MQ'
+    | 'MR'
+    | 'MU'
+    | 'YT'
+    | 'MX'
+    | 'FM'
+    | 'MD'
+    | 'MC'
+    | 'MN'
+    | 'ME'
+    | 'MS'
+    | 'MA'
+    | 'MZ'
+    | 'MM'
+    | 'NA'
+    | 'NR'
+    | 'NP'
+    | 'NL'
+    | 'NC'
+    | 'NZ'
+    | 'NI'
+    | 'NE'
+    | 'NG'
+    | 'NU'
+    | 'NF'
+    | 'MP'
+    | 'NO'
+    | 'OM'
+    | 'PK'
+    | 'PW'
+    | 'PS'
+    | 'PA'
+    | 'PG'
+    | 'PY'
+    | 'PE'
+    | 'PH'
+    | 'PN'
+    | 'PL'
+    | 'PT'
+    | 'PR'
+    | 'QA'
+    | 'RE'
+    | 'RO'
+    | 'RU'
+    | 'RW'
+    | 'BL'
+    | 'SH'
+    | 'KN'
+    | 'LC'
+    | 'MF'
+    | 'PM'
+    | 'VC'
+    | 'WS'
+    | 'SM'
+    | 'ST'
+    | 'SA'
+    | 'SN'
+    | 'RS'
+    | 'SC'
+    | 'SL'
+    | 'SG'
+    | 'SX'
+    | 'SK'
+    | 'SI'
+    | 'SB'
+    | 'SO'
+    | 'ZA'
+    | 'GS'
+    | 'SS'
+    | 'ES'
+    | 'LK'
+    | 'SD'
+    | 'SR'
+    | 'SJ'
+    | 'SZ'
+    | 'SE'
+    | 'CH'
+    | 'SY'
+    | 'TW'
+    | 'TJ'
+    | 'TZ'
+    | 'TH'
+    | 'TL'
+    | 'TG'
+    | 'TK'
+    | 'TO'
+    | 'TT'
+    | 'TN'
+    | 'TR'
+    | 'TM'
+    | 'TC'
+    | 'TV'
+    | 'UG'
+    | 'UA'
+    | 'AE'
+    | 'GB'
+    | 'US'
+    | 'UM'
+    | 'UY'
+    | 'UZ'
+    | 'VU'
+    | 'VE'
+    | 'VN'
+    | 'VG'
+    | 'VI'
+    | 'WF'
+    | 'EH'
+    | 'YE'
+    | 'ZM'
+    | 'ZW'
+
+const regionNameMap = new Map(regions.map((r) => [r.code, r.name])) as Map<RegionCode, string>
+
+type Region = {
+    code: RegionCode
+    name: string
+}
+
+interface RegionResolver {
+    getRegion(): Promise<Region>
+}
+
+function createRegionResolver(api: string, field: string): RegionResolver {
+    return {
+        getRegion: async (): Promise<Region> => {
+            const response = await fetch(api)
+            const result = await response.json()
+            const code = result[field]
+            if (!code) {
+                throw new Error('Failed to resolve region code')
+            }
+            return {
+                code,
+                name: regionNameMap.get(code)!,
+            }
+        },
+    }
+}
+
+const IPGeoLocationAPI = createRegionResolver('https://api.ipgeolocationapi.com/geolocate', 'alpha2')
+// TODO the api https endpoint needs be paid
+const IPAPI = createRegionResolver('https://ip-api.com/json/?fields=status,countryCode', 'countryCode')
+
+export function useRegion(): AsyncState<Region> {
+    const { loading, error, value } = useAsync(IPGeoLocationAPI.getRegion)
```

Comment:
`useRegion` should return an `AsyncResult<Regoin>`.

Reply:
I searched the maskbook and react-use, can't find anywhere fined `AsyncResult`

Reply:
I searched the maskbook and react-use, can't find anywhere fined `AsyncResult`

Reply:
```
export function useRegion() {
    return useAsync(...)
}
```

Reply:
Sorry, It's `AsyncStateRetry<Regoin>`.